runtime/src/main/java/org/corfudb/runtime/collections/ISMRMap.java

```diff
@@ -272,4 +273,55 @@ default void undoPutAll(ISMRMap<K,V> map, Map<K,V> undoRecord,
     @Override
     Set<Entry<K, V>> entrySet();
 
+    @DontInstrument
+    Object getMapGarbageManager();
+
+    /**
+     * This class is used to maintain the location and other information on different keys.
+     * @param <T> the type of the map key.
+     */
+    class SMRMapGarbageManager<T>{
+
+        // Keys have 1:1 mapping to SMRRecord, which includes the location info.
+        // The location of each key should be increase monotonically.
+        final private Map<T, SMRRecord.SMRRecordInfo> keyToSMRRecordInfo = new ConcurrentHashMap<>();
+
+        private SMRRecordLocator last = SMRRecordLocator.NONE;
+
+        public Optional<SMRRecord.SMRRecordInfo> addUnsafe(@NonNull T key,
+                                                           @NonNull SMRRecord.SMRRecordInfo smrRecordInfo) {
+            SMRRecordLocator locator = smrRecordInfo.getLocator();
+            if (locator.compareTo(last) <= 0) {
```

Comment:
I remember last time we discussed this should be done in VLO to insure the address does not regress, right?

Reply:
I am sorry I forget this argument. Could we talk about this issue offline?

Reply:
I think we could put the logic in either SMRMapGarbageManger or  VLO.